pkg/sriov/networkservice/mechanisms/vfio/client.go

```
@@ -0,0 +1,81 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vfio
+
+import (
+	"context"
+	"os"
+	"path"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+	"golang.org/x/sys/unix"
+	"google.golang.org/grpc"
+)
+
+type vfioClient struct {
+	vfioDir   string
+	cgroupDir string
+}
+
+// NewClient returns a new VFIO client chain element
+func NewClient(vfioDir, cgroupDir string) networkservice.NetworkServiceClient {
+	return &vfioClient{
+		vfioDir:   vfioDir,
+		cgroupDir: cgroupDir,
+	}
+}
+
+func (c *vfioClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	logEntry := log.Entry(ctx).WithField("vfioClient", "Request")
+
+	request.Connection.Context.ExtraContext[clientCgroupDirKey] = c.cgroupDir
+
+	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	_ = os.Mkdir(c.vfioDir, 0600)
```

Interesting... Why do we ignore the error here? Also, can we move 0600 to const?

---

I was thinking of case when we fail on double creating the VFIO directory, but there is an `os.IsExist(err)`...
So fixed :)